cpp/include/cuspatial/experimental/detail/linestring_intersection_count.cuh

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/detail/utility/device_atomics.cuh>
+#include <cuspatial/detail/utility/linestring.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/tuple.h>
+
+namespace cuspatial {
+namespace detail {
+
+template <typename MultiLinestringRange1,
+          typename MultiLinestringRange2,
+          typename OutputIt1,
+          typename OutputIt2>
+__global__ void count_intersection_and_overlaps_simple(MultiLinestringRange1 multilinestrings1,
+                                                       MultiLinestringRange2 multilinestrings2,
+                                                       OutputIt1 point_count_it,
+                                                       OutputIt2 segment_count_it)
+{
+  using T = typename MultiLinestringRange1::element_t;
+  for (auto idx = threadIdx.x + blockIdx.x * blockDim.x; idx < multilinestrings1.num_points();
+       idx += gridDim.x * blockDim.x) {
+    auto const part_idx = multilinestrings1.part_idx_from_point_idx(idx);
+    if (!multilinestrings1.is_valid_segment_id(idx, part_idx)) continue;
+    auto const geometry_idx = multilinestrings1.geometry_idx_from_part_idx(part_idx);
+    auto [a, b]             = multilinestrings1.segment(idx);
+
+    for (auto const& linestring2 : multilinestrings2[geometry_idx]) {
+      for (auto [c, d] : linestring2) {
+        auto [point_opt, segment_opt] = segment_intersection(segment<T>{a, b}, segment<T>{c, d});
+        if (point_opt.has_value()) {
+          auto r = make_atomic_ref(*(point_count_it + geometry_idx));
+          r.fetch_add(1);
```

Comment:
The `MEMBAR` comes from the fact that by default, `fetch_add` uses `memory_order_seqcst`, which is the strongest memory ordering and likely not what you want. 

Unless you are using the atomic operation as a signaling mechanism between threads, you likely just want to use the weakest memory ordering, `memory_order_relaxed`.

https://godbolt.org/z/8h1nsczE7

More info: https://www.youtube.com/watch?v=A8eCGOqgvH4
```suggestion
          r.fetch_add(1, cuda::memory_order_relaxed);
```